modules/videoio/test/test_audiocombination.cpp

```diff
@@ -0,0 +1,150 @@
+#include "test_precomp.hpp"
+
+#include <iostream>
+namespace opencv_test { namespace {
+
+//file format, epsilon, video type, weight, height, number of frame, fps, psnr Threshold, backend
+typedef std::tuple<std::string, double, int, int, int, int, int, double, std::pair<std::string, int> > Param;
+
+const Param audio_params[] =
+{
+    Param("mp4", 0.15, CV_8UC3, 240, 320, 90, 30, 30., {"CAP_MSMF", cv::CAP_MSMF})
+};
+
+class AudioTestFixture : public testing::TestWithParam <Param>
+{
+public:
+    AudioTestFixture():
+        format(get<0>(GetParam())),
+        epsilon(get<1>(GetParam())),
+        videoType(get<2>(GetParam())),
+        height(get<3>(GetParam())),
+        width(get<4>(GetParam())),
+        numberOfFrames(get<5>(GetParam())),
+        fps(get<6>(GetParam())),
+        psnrThreshold(get<7>(GetParam())),
+        backend(get<8>(GetParam())),
+        root("audio/"),
+        fileName("test_audio"),
+        params({ CAP_PROP_AUDIO_STREAM, 0,
+                 CAP_PROP_VIDEO_STREAM, 0,
+                 CAP_PROP_AUDIO_DATA_DEPTH, CV_16S  }) {};
+    void doTest()
+    {
+        getValidAudioData();
+        getValidVideoData();
+        getFileData();
+        comparisonAudio();
+        comparisonVideo();
+    }
+
+private:
+    void getValidAudioData()
+    {
+        const double step = 3.14/22050;
+        double value = 0;
+        for (int j = 0; j < 3; j++)
+        {
+            value = 0;
+            for (int i = 0; i < 44100; i++)
+            {
+                validAudioData.push_back(sin(value));
+                value += step;
+            }
+        }
+    }
+    void getValidVideoData()
+    {
+        Mat img(height, width, videoType);
+        for (int i = 0; i < numberOfFrames; ++i)
+        {
+            generateFrame(i, numberOfFrames, img);
+            validVideoData.push_back(img);
+        }
+    }
+    void getFileData()
+    {
+        ASSERT_TRUE(cap.open(findDataFile(root + fileName + "." + format), backend.second, params));
+        const int audioBaseIndex = static_cast<int>(cap.get(cv::CAP_PROP_AUDIO_BASE_INDEX));
+        double f = 0;
+        for (;;)
+        {
+            if (cap.grab())
+            {
+                ASSERT_TRUE(cap.retrieve(videoFrame));
+                ASSERT_TRUE(cap.retrieve(audioFrame, audioBaseIndex));
+                ASSERT_EQ(audioFrame.cols/44100, 1/fps); // check if the duration of the received audio data satisfies one video frame
```

Comment:
Does this check may fail?

Reply:
Yes. If the duration of the received audio data does not equal to one frame of the video, the check will not be passed. This is possible if synchronization does not work

Reply:
Provided code doesn't look valid. Looks like it always compare 0 vs 0.

Use this instead:

int samplesPerFrame = 1/fps*44100;
// 44100 should be checked and replaced to `CAP_PROP_AUDIO_SAMPLES_PER_SECOND` property.

int audioSamplesTolerance = samplesPerFrame / 2;

We should have these checks:
- position: abs(CAP_PROP_AUDIO_POS - CAP_PROP_POS_MSEC / 1000 * 44100) < audioSamplesTolerance
- duration: abs(audioFrame.cols - samplesPerFrame) < audioSamplesTolerance

Reply:
> Provided code doesn't look valid. Looks like it always compare 0 vs 0.
> 
> Use this instead:
> 
> int samplesPerFrame = 1/fps*44100;
> // 44100 should be checked and replaced to `CAP_PROP_AUDIO_SAMPLES_PER_SECOND` property.
> 
> int audioSamplesTolerance = samplesPerFrame / 2;
> 
> We should have these checks:
> 
> * position: abs(CAP_PROP_AUDIO_POS - CAP_PROP_POS_MSEC / 1000 * 44100) < audioSamplesTolerance
> * duration: abs(audioFrame.cols - samplesPerFrame) < audioSamplesTolerance

The CAP_PROP_POS_MSEC value is not always valid